### ISSUE TITLE
Unexpected memory usage in FSDP 2 Hybrid Sharding (HSDP)

### ISSUE THREAD

**Issue by Craigacp** [pytorch/pytorch#151030]
### 🐛 Describe the bug

FSDP 2 & torch.distributed.checkpoint has unexpected additional memory usage when saving, which caused our jobs to fail. I'm not clear if this is something I've misconfigured (our stack is built on pytorch & HuggingFace but we're not using HF's trainer or accelerate), or if this is intended behaviour from the checkpointing system. 

We are doing a 64 GPU training run on 8 OCI BM.GPU.H100.8 machines, with sharding across the 8 GPUs in a node and data parallel replicas across the 8 machines, training a ~1B parameter transformer decoder. We ended up on hybrid sharding due to some external networking factors though it's not much faster than full sharding in our setup. GPU 0 has 7 extra processes running which steady state at 520MB VRAM, and when we checkpoint the model using `torch.distributed.checkpoint` they jump up to 2+GB each, for a total of 11+GB of additional VRAM. My understanding of how the checkpoint system works was that there is no communication between the GPUs when saving out a checkpoint as it is saved by each shard independently.

It looks like these additional processes aren't using the same pytorch memory allocator, as we checkpoint after gradients & activations have been cleared, so while there is lots of VRAM used it's all free in the caching allocator. We tested this by clearing the cache before checkpointing and the checkpoint is fine. Oddly the run crashes with OOM on the next immediate step, presumably due to some kind of fragmentation issue as all our batches are the same size.

The relevant code for sharding & checkpoint saving is excerpted below. I think we might be getting the state dict out of the model incorrectly, though I'm still unclear what the additional processes are for.

```python
def wrap_model(self, comp_env):
    # comp_env is a data class which contains the job configuration from SLURM
    # this method is on a data class which contains the FSDP 2 configuration
    num_nodes = comp_env.world_size // comp_env.node_size
    self.mesh = init_device_mesh(device_type=comp_env.device.type, mesh_shape=(num_nodes, comp_env.node_size), mesh_dim_names=("dp", "shard"))
    config = {
        "mp_policy": self.mixed_precision,
        "offload_policy": self.offload_policy,
        "mesh": self.mesh,
        "reshard_after_forward": True,  # when true implements full sharding, when false implements ZeRO-2
    }
    if self.mixed_precision:
        top_level_mixed_precision = MixedPrecisionPolicy(param_dtype=self.mixed_precision.param_dtype,
                                                         reduce_dtype=self.mixed_precision.reduce_dtype,
                                                         output_dtype=torch.float32,
                                                         cast_forward_inputs=self.mixed_precision.cast_forward_inputs)
    else:
        top_level_mixed_precision = None

    cls_list_to_wrap = self._transformer_cls(model)

    def wrap_policy(module: torch.nn.Module) -> bool:
        if self.layer_cls_to_wrap is None:
            return False
        return isinstance(module, tuple(cls_list_to_wrap))

    model.compile() # does layer wise compilation like torchtitan's llama

    stack = [model]
    ordered_modules = []
    while stack:
        current_module = stack.pop()
        for _, attr in current_module.named_children():
            if isinstance(attr, torch.nn.Module):
                stack.append(attr)
        ordered_modules.append(current_module)

    # wrap all the modules which match the policy aside from the top level module
    for module in ordered_modules[::-1][:-1]:
        if wrap_policy(module):
            fully_shard(module, **config)

    # wrap the top level module with a different MP config so it always emits fp32 outputs
    top_level_config = config.copy()
    top_level_config['mp_policy'] = top_level_mixed_precision
    fully_shard(model, **top_level_config)

    return model

def checkpoint_model_and_optimizer(self, comp_env, model, optimizer, output_dir):
    state_dict = model.state_dict()

    ckpt_dir = os.path.join(output_dir, f"{MODEL_NAME}")
    os.makedirs(ckpt_dir, exist_ok=True)
    logger.info(f"Saving model to {ckpt_dir}")
    state_dict = {"model": state_dict}
    # Only save the first replica
    if comp_env.node_rank == 0:
        dist_ckpt.save(
            state_dict=state_dict,
            storage_writer=dist_ckpt.FileSystemWriter(ckpt_dir),
            process_group=self.mesh.get_group("shard")
        )
    comp_env.wait_for_everyone()
    logger.info(f"Model saved to {ckpt_dir}")

    from torch.distributed.checkpoint.state_dict import get_optimizer_state_dict, StateDictOptions
    optim_state = get_optimizer_state_dict(model, optimizer, options=StateDictOptions(flatten_optimizer_state_dict=True))
    ckpt_dir = os.path.join(output_dir, f"{OPTIMIZER_NAME}")
    os.makedirs(ckpt_dir, exist_ok=True)
    # Only save the first replica
    if comp_env.node_rank == 0:
        dist_ckpt.save(
            state_dict={"optimizer": optim_state},
            storage_writer=dist_ckpt.FileSystemWriter(ckpt_dir),
            planner=DefaultSavePlanner(),
            process_group=self.mesh.get_group("shard")
        )
    comp_env.wait_for_everyone()
    logger.info(f"Optimizer state saved in {ckpt_dir}")
```

If this code is using FSDP 2 & dcp correctly, then what is causing the memory bump and is there any way to predict how big it will be? Alternatively if I've got some of this wrong then is there a source beyond looking at the torchtitan code for how to do hybrid sharding correctly? We didn't hit this issue when doing full sharding across the 64 GPUs, though that obviously had slightly lower memory pressure so we might have just been lucky.

### Versions

Collecting environment information...
PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Oracle Linux Server 8.10 (x86_64)
GCC version: (GCC) 8.5.0 20210514 (Red Hat 8.5.0-24.0.1)
Clang version: Could not collect
CMake version: version 3.26.5
Libc version: glibc-2.28

Python version: 3.12.7 (main, Mar 29 2025, 19:03:01) [GCC 8.5.0 20210514 (Red Hat 8.5.0-23.0.1)] (64-bit runtime)
Python platform: Linux-4.18.0-553.27.1.el8_10.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3
GPU 2: NVIDIA H100 80GB HBM3
GPU 3: NVIDIA H100 80GB HBM3
GPU 4: NVIDIA H100 80GB HBM3
GPU 5: NVIDIA H100 80GB HBM3
GPU 6: NVIDIA H100 80GB HBM3
GPU 7: NVIDIA H100 80GB HBM3

Nvidia driver version: 560.35.05
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.8.9.7
/usr/lib64/libcudnn_adv_infer.so.8.9.7
/usr/lib64/libcudnn_adv_train.so.8.9.7
/usr/lib64/libcudnn_cnn_infer.so.8.9.7
/usr/lib64/libcudnn_cnn_train.so.8.9.7
/usr/lib64/libcudnn_ops_infer.so.8.9.7
/usr/lib64/libcudnn_ops_train.so.8.9.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:         x86_64
CPU op-mode(s):       32-bit, 64-bit
Byte Order:           Little Endian
CPU(s):               224
On-line CPU(s) list:  0-111
Off-line CPU(s) list: 112-223
Thread(s) per core:   1
Core(s) per socket:   56
Socket(s):            2
NUMA node(s):         2
Vendor ID:            GenuineIntel
CPU family:           6
Model:                143
Model name:           Intel(R) Xeon(R) Platinum 8480+
Stepping:             8
CPU MHz:              2000.000
CPU max MHz:          3800.0000
CPU min MHz:          800.0000
BogoMIPS:             4000.00
Virtualization:       VT-x
L1d cache:            48K
L1i cache:            32K
L2 cache:             2048K
L3 cache:             107520K
NUMA node0 CPU(s):    0-55
NUMA node1 CPU(s):    56-111
Flags:                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] mypy==1.11.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.2.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] torch==2.6.0
[pip3] torch-tb-profiler==0.4.3
[pip3] torchvision==0.21.0
[pip3] triton==3.2.0
[conda] Could not collect

**Comment by fegin** [comment]
@Craigacp Can you try https://docs.pytorch.org/docs/stable/torch_cuda_memory.html to understand what the extra memory is from? My understanding is that FSDP2 state_dict() shouldn't incur extra memory usage.

**Comment by Craigacp** [comment]
Will that track memory used in different processes, or is it just for the main Python process?

**Comment by fegin** [comment]
What does this mean? Are you using multiple PyThon processes on a single rank?  Or are you using multiple threads. That has different implications. The memory tracker will only track the called PyThon process. All other memory usage may be shown as non-trackable memory usage. Even if it can detect that memory usage, the memory tracker won't know the call stack because it locates on a different process.
